### PR TITLE
Convert byte_extract keyword parser to Rust

### PR DESCRIPTION
**Pull request by jlucovsky**
Continuation of #10833 

Convert the byte_extract option parser from C to Rust.

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: [6873](https://redmine.openinfosecfoundation.org/issues/6873)

Describe changes:
- Refactor code in rust/src/detect to support re-usability
- Implement the byte_extract parser in Rust, with unittests
- Modify the existing byte_extract module to use the Rust parser

Updates:
- Rebase

### Provide values to any of the below to override the defaults.

To use a pull request use a branch name like `pr/N` where `N` is the
pull request number.

Alternatively, `SV_BRANCH` may also be a link to an
OISF/suricata-verify pull-request.

```
SV_REPO=
SV_BRANCH=
SU_REPO=
SU_BRANCH=
LIBHTP_REPO=
LIBHTP_BRANCH=
```

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 20129

**Review by catenacyber — CHANGES_REQUESTED** [review]
Thanks for the work Jeff

CI : 🟢 
Code : Checking now
Commits segmentation : Could last commit get squashed ?
Commit messages : The redmine ticket 6873 should be mentioned in the commit detect/byte_extract: Move keyword parser to Rust
Git ID set : looks fine for me
CLA : :-p
Doc update : not needed
Redmine ticket : Is https://redmine.openinfosecfoundation.org/issues/6831 handled here ?
Rustfmt : Could you run it to have 2 small changes in byte_extract.rs and byte_math.rs ?
Tests : ok
Dependencies added: none added

File: rust/src/detect/byte_math.rs [hunk 1]
```diff
@@ -491,8 +453,18 @@ mod tests {
     }
 
     fn valid_test(
-        args: &str, nbytes: u8, offset: i32, oper: ByteMathOperator, rvalue_str: &str, nbytes_str: &str, rvalue: u32,
```
Comment (catenacyber) [inline rust/src/detect/byte_math.rs:494]: rustfmt does not agree with this change 🤔

File: rust/src/detect/byte_math.rs [hunk 2]
```diff
@@ -298,8 +260,8 @@ fn parse_bytemath(input: &str) -> IResult<&str, DetectByteMathData, RuleParseErr
                     return Err(make_error("endianess already set".to_string()));
                 }
                 byte_math.endian = match get_endian_value(val) {
```
[the remark this answers is not in the document]
Comment (jlucovsky) [reply rust/src/detect/byte_math.rs:262]: I'll make this change for `endian` and `base`

File: rust/src/detect/byte_math.rs (hunk 1 above)
Comment (jlucovsky) [reply rust/src/detect/byte_math.rs:494]: @catenacyber I reran `rustfmt` and it yielded the same results. What are you seeing that shows `rustfmt` doesn't like the change?

Comment (catenacyber) [reply rust/src/detect/byte_math.rs:494]:
```diff
diff --git a/rust/src/detect/byte_extract.rs b/rust/src/detect/byte_extract.rs
index 6839d19c57..51bd6b78ef 100644
--- a/rust/src/detect/byte_extract.rs
+++ b/rust/src/detect/byte_extract.rs
@@ -319,15 +319,8 @@ mod tests {
     }
 
     fn valid_test(
-        args: &str,
-        nbytes: u8,
-        offset: i16,
-        var_name_str: &str,
-        base: ByteBase,
-        endian: ByteEndian,
-        align_value: u8,
-        multiplier_value: u16,
-        flags: u16,
+        args: &str, nbytes: u8, offset: i16, var_name_str: &str, base: ByteBase,
+        endian: ByteEndian, align_value: u8, multiplier_value: u16, flags: u16,
     ) {
         let bed = SCDetectByteExtractData {
             nbytes,
diff --git a/rust/src/detect/byte_math.rs b/rust/src/detect/byte_math.rs
index 2b29c5b26e..4d489762f2 100644
--- a/rust/src/detect/byte_math.rs
+++ b/rust/src/detect/byte_math.rs
@@ -453,18 +453,9 @@ mod tests {
     }
 
     fn valid_test(
-        args: &str,
-        nbytes: u8,
-        offset: i32,
-        oper: ByteMathOperator,
-        rvalue_str: &str,
-        nbytes_str: &str,
-        rvalue: u32,
-        result: &str,
-        base: ByteBase,
-        endian: ByteEndian,
-        bitmask_val: u32,
-        flags: u8,
+        args: &str, nbytes: u8, offset: i32, oper: ByteMathOperator, rvalue_str: &str,
+        nbytes_str: &str, rvalue: u32, result: &str, base: ByteBase, endian: ByteEndian,
+        bitmask_val: u32, flags: u8,
     ) {
         let bmd = DetectByteMathData {
             nbytes,
```

Comment (catenacyber) [reply rust/src/detect/byte_math.rs:494]: With rustfmt 1.7.0-nightly (1388d7a0 2024-03-20)

Comment (jasonish) [reply rust/src/detect/byte_math.rs:494]: We should probably use latest Rust stable for fmt checks.

Comment (jlucovsky) [reply rust/src/detect/byte_math.rs:494]: @catenacyber Thanks ... I was using an older version of `rustfmt` -- 1.5

**Comment by jlucovsky** [comment]
Continued in #10906

**Pull request closed without merging**